pkg/gadgets/run/tracer/tracer.go

```diff
@@ -177,6 +179,136 @@ func (t *Tracer) Close() {
 	}
 }
 
+var (
+	once             sync.Once
+	ringbufAvailable bool
+
+	// Prefix used to mark tracer map created with GADGET_TRACER_MAP().
+	tracerMapPrefix = "gadget_tracer_map_"
+)
+
+func isRingbufAvailable() bool {
+	once.Do(func() {
+		ringbuf, err := ebpf.NewMap(&ebpf.MapSpec{
+			Type:       ebpf.RingBuf,
+			MaxEntries: uint32(os.Getpagesize()),
+		})
+
+		ringbuf.Close()
+
+		ringbufAvailable = err == nil
+	})
+
+	return ringbufAvailable
+}
+
+// createdByTracerMapMacro returns whether the tracer map was created using
+// GADGET_TRACER_MAP().
+func (t *Tracer) createdByTracerMapMacro(tracerMapName string) bool {
+	results, err := types.GetGadgetIdentByPrefix(t.spec, tracerMapPrefix)
+	if err != nil {
+		return false
+	}
+
+	for _, result := range results {
+		if result == tracerMapName {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (t *Tracer) handleTracerMapDefinition(tracerMapName string) error {
+	if !isRingbufAvailable() {
+		log.Debugf("Ring buffers are not available, defaulting to perf ones")
+		if tracerMapName != "" {
+			bufMap, ok := t.spec.Maps[tracerMapName]
+			if !ok {
+				return fmt.Errorf("no buffer map named %s", tracerMapName)
+			}
+
+			bufMap.Type = ebpf.PerfEventArray
+			bufMap.KeySize = 4
+			bufMap.ValueSize = 4
+
+			t.spec.Maps[tracerMapName] = bufMap
+		}
+	} else {
+		heapMapName := "gadget_heap"
+		// If we delete the gadget_heap map, the verifier will not load the code
+		// because it cannot find it despite the code using it being dead one:
+		// ...: create BPF collection: program ig_mount_x: missing map gadget_heap
+		// Rather than deleting it, we just set its size to 0 and use a hash to
+		// avoid having one struct per CPU, i.e. reducing as much as possible the
+		// memory footprint it uses.
+		heapMap, ok := t.spec.Maps[heapMapName]
+		if ok {
+			heapMap.Type = ebpf.Hash
+			heapMap.ValueSize = 4
+
+			t.spec.Maps[heapMapName] = heapMap
+		}
+	}
+
+	return nil
+}
+
+func (t *Tracer) createTracerMap(tracerMapName string) error {
+	var err error
+
+	if !isRingbufAvailable() {
+		t.perfReader, err = perf.NewReader(t.collection.Maps[tracerMapName], gadgets.PerfBufferPages*os.Getpagesize())
+
+		return err
+	}
+
+	t.ringbufReader, err = ringbuf.NewReader(t.collection.Maps[tracerMapName])
+
+	return err
+}
+
+func (t *Tracer) loadeBPFObjects(opts ebpf.CollectionOptions, tracerMapName string) error {
```

Comment:
I like having this small functions to make the code more modular. I suspect in the future we'll be adding more parameters to it, so what about using a structure as parameters from the beginning?